### ISSUE TITLE
GRPOTrainer training causes model to completely deteriorate when training with accelerate

### ISSUE THREAD

**Issue by Randomdude11** [huggingface/trl#3881]
Hello,

I have been trying to speed up my training on GRPOTrainer with vllm + accelerate on 8 GPUs, but strangely the same script which used to train perfectly fine (albeit more slowly) without vllm or accelerate, causes the model's performance to drastically deteriorate when I use them. I removed vllm and realized that even with accelerate only, the problem persists. When I remove accelerate altogether and run the script with python, training runs normally.

The only difference is that without accelerate I load the model with device_map = 'auto'.

I tried playing around with batch sizes (I know with accelerate the batch size is multiplied by the number of devices), tried pre-tokenizing, passing a processing_class, I tried passing a model_id instead of a pre-loaded model with or without processing classes, pre-loading the model/tokenizer with or without trust_remote_code=True, pretty much everything. Still have no idea why using accelerate to train causes the model to completely deteriorate.

Would appreciate any feedback or ideas, because I am completely at a loss.

Thanks in advance!

**Comment by Randomdude11** [comment]
I just found that downgrading to transformers==4.51.3 (and therefore to trl==0.19.0) makes training with accelerate work again. What breaking change could have caused this issue?

**Comment by qgallouedec** [comment]
Thanks for the explanation. A minimal reproducer would actually help

**Comment by Randomdude11** [comment]
Thanks for getting back. Here is a minimal script which should reproduce this behavior. Unfortunately I cannot verify this because GPUs in my cluster are very busy, but I have stripped parts of the code which very likely do not influence trl/transformers/accelerate:

```
import torch
from datasets import Dataset, Features, Value
from transformers import AutoTokenizer, AutoModelForCausalLM
from trl import (
    GRPOConfig,
    GRPOTrainer,
)
import json
import my_rewards as rewards

def make_dp(data_dict):
    return {
        "prompt": data_dict['text'],
        "solution": data_dict['label']
    }

if __name__ == "__main__":
    ################
    # Model
    ################
    model_id = "deepseek-ai/DeepSeek-R1-0528-Qwen3-8B"
    model = AutoModelForCausalLM.from_pretrained(
        model_id,
        torch_dtype=torch.bfloat16,
        low_cpu_mem_usage=True,
        trust_remote_code=True,
        device_map='auto',
    )
    tokenizer = AutoTokenizer.from_pretrained(model_id, trust_remote_code=True)
    ################
    # Dataset
    ################
    with open('train_raw.json', 'r') as f:
        raw_dict = json.load(f)
    train_data = [make_dp(rd)
    for rd in raw_dict.values()
    ]
    with open('eval_raw.json', 'r') as f:
        raw_dict = json.load(f)
    eval_data = [make_dp(rd)
    for rd in raw_dict.values()
    ]
    features = Features({
        "prompt": [{
            "role": Value("string"),
            "content": Value("string")
        }],
        "solution": Value("int32"),
    })
    train_dataset = Dataset.from_list(train_data, features=features) 
    eval_dataset = Dataset.from_list(eval_data, features=features)
    ################
    # Training
    ################
    config = GRPOConfig(
        output_dir="checkpoint_dir",
        learning_rate=1e-5,
        num_train_epochs=100,
        num_generations=4,
        per_device_train_batch_size=16,
        gradient_accumulation_steps=2,
        max_prompt_length=1024,
        max_completion_length=2048,
        bf16=True,
        logging_steps=5,
        eval_strategy='steps',
        eval_steps=750,
        save_strategy='steps',
        save_steps=750,
        log_completions=True,
        disable_tqdm=False,
        logging_strategy='steps',
        logging_dir='checkpoint_dir',
        reward_weights=[0.7, 0.3, 0.0, 0.0, 0.0, 0.0],
        run_name='my_run'
    )
    trainer = GRPOTrainer(
        model=model,
        processing_class=tokenizer,
        args=config,
        reward_funcs=[rewards.accuracy_reward, rewards.is_final_answer_format_reward, rewards.absolute_error, rewards.squared_error, rewards.final_answer, rewards.ground_truth],
        train_dataset=train_dataset,
        eval_dataset=eval_dataset
    )

    trainer.train()
```

I also tried tweaking many things in the script or my accelerate config, but let me know if anything else would make this easier.

**Comment by qgallouedec** [comment]
Hey @Randomdude11, sorry for the slow reply, catching up on the backlog.

A couple of things make this hard to action:
- the script imports a custom `my_rewards` module and reads `train_raw.json`/`eval_raw.json` that aren't shared, and you noted you couldn't actually verify the snippet runs end-to-end, so we don't have a reproducer we can run;
- there's no `trl env` output, so we don't know exactly which versions you saw the regression at.

One thing that stands out independent of the regression: `AutoModelForCausalLM.from_pretrained(..., device_map='auto')` should generally not be combined with `accelerate launch`: when launched on N processes each rank will try to auto-shard a separate copy of the model across all GPUs, which can absolutely lead to broken training. With DDP under accelerate you want to drop `device_map` and let accelerate place the model. Closing as won't solve for now; happy to re-open if you can share a self-contained repro
